### DISCUSSION TITLE
Uncaught exception with Fabric Loader

### DISCUSSION THREAD

**Discussion by Frankenberrypi** [itzg/docker-minecraft-server#2565] [Q&A] [answered]
This is likely user error since it's my first time trying this.  I'm able to get a server running just fine with vanilla, but get errors when trying to use Fabric.  My ultimate goal is to run mods to support the Hermitcraft S9 map download, but currently I'm just trying to get Fabric running on its own.  

It looks like the error in question is:
```
[FabricLoader/]: Uncaught exception in thread "main"
```

## Docker compose file  
```
version: "3.8"

services:
  hermitcraft9:
    image: itzg/minecraft-server
    tty: true
    restart: always
    stdin_open: true
    ports:
      - 25565:25565
    environment:
      EULA: TRUE
      VERSION: 1.20.1
      TYPE: FABRIC
    volumes:
      # attach the relative directory 'data' to the container's /data path
      - /mnt/network_storage/Minecraft_Worlds_Java/hc_tester:/data
```

## Logs  
```
[init] Changing ownership of /data to 1000 ...

[init] Running as uid=1000 gid=1000 with /data as 'drwxrwxrwx 2 0 0 0 Dec 30 16:14 /data'

[init] Resolving type given FABRIC

[mc-image-helper] 16:14:04.291 INFO  : Downloaded ./fabric-server-mc.1.20.1-loader.0.15.3-launcher.0.11.2.jar

[init] Creating server properties in /data/server.properties

[init] Disabling whitelist functionality

[mc-image-helper] 16:14:06.693 INFO  : Created/updated 4 properties in /data/server.properties

[init] Setting initial memory to 1G and max to 1G

[init] Starting the Minecraft server...

Downloading Minecraft server

Installing Fabric Loader 0.15.3(1.20.1) on the server

Downloading required files

Downloading library org.ow2.asm:asm:9.6

Downloading library org.ow2.asm:asm-analysis:9.6

Downloading library org.ow2.asm:asm-commons:9.6

Downloading library org.ow2.asm:asm-tree:9.6

Downloading library org.ow2.asm:asm-util:9.6

Downloading library net.fabricmc:sponge-mixin:0.12.5+mixin.0.8.5

Downloading library net.fabricmc:intermediary:1.20.1

Downloading library net.fabricmc:fabric-loader:0.15.3

Generating server launch JAR

Unpacking 1.20.1/server-1.20.1.jar (versions:1.20.1) to versions/1.20.1/server-1.20.1.jar

Unpacking com/github/oshi/oshi-core/6.2.2/oshi-core-6.2.2.jar (libraries:com.github.oshi:oshi-core:6.2.2) to libraries/com/github/oshi/oshi-core/6.2.2/oshi-core-6.2.2.jar

Unpacking com/google/code/gson/gson/2.10/gson-2.10.jar (libraries:com.google.code.gson:gson:2.10) to libraries/com/google/code/gson/gson/2.10/gson-2.10.jar

Unpacking com/google/guava/failureaccess/1.0.1/failureaccess-1.0.1.jar (libraries:com.google.guava:failureaccess:1.0.1) to libraries/com/google/guava/failureaccess/1.0.1/failureaccess-1.0.1.jar

Unpacking com/google/guava/guava/31.1-jre/guava-31.1-jre.jar (libraries:com.google.guava:guava:31.1-jre) to libraries/com/google/guava/guava/31.1-jre/guava-31.1-jre.jar

Unpacking com/mojang/authlib/4.0.43/authlib-4.0.43.jar (libraries:com.mojang:authlib:4.0.43) to libraries/com/mojang/authlib/4.0.43/authlib-4.0.43.jar

Unpacking com/mojang/brigadier/1.1.8/brigadier-1.1.8.jar (libraries:com.mojang:brigadier:1.1.8) to libraries/com/mojang/brigadier/1.1.8/brigadier-1.1.8.jar

Unpacking com/mojang/datafixerupper/6.0.8/datafixerupper-6.0.8.jar (libraries:com.mojang:datafixerupper:6.0.8) to libraries/com/mojang/datafixerupper/6.0.8/datafixerupper-6.0.8.jar

Unpacking com/mojang/logging/1.1.1/logging-1.1.1.jar (libraries:com.mojang:logging:1.1.1) to libraries/com/mojang/logging/1.1.1/logging-1.1.1.jar

Unpacking commons-io/commons-io/2.11.0/commons-io-2.11.0.jar (libraries:commons-io:commons-io:2.11.0) to libraries/commons-io/commons-io/2.11.0/commons-io-2.11.0.jar

Unpacking io/netty/netty-buffer/4.1.82.Final/netty-buffer-4.1.82.Final.jar (libraries:io.netty:netty-buffer:4.1.82.Final) to libraries/io/netty/netty-buffer/4.1.82.Final/netty-buffer-4.1.82.Final.jar

Unpacking io/netty/netty-codec/4.1.82.Final/netty-codec-4.1.82.Final.jar (libraries:io.netty:netty-codec:4.1.82.Final) to libraries/io/netty/netty-codec/4.1.82.Final/netty-codec-4.1.82.Final.jar

Unpacking io/netty/netty-common/4.1.82.Final/netty-common-4.1.82.Final.jar (libraries:io.netty:netty-common:4.1.82.Final) to libraries/io/netty/netty-common/4.1.82.Final/netty-common-4.1.82.Final.jar

Unpacking io/netty/netty-handler/4.1.82.Final/netty-handler-4.1.82.Final.jar (libraries:io.netty:netty-handler:4.1.82.Final) to libraries/io/netty/netty-handler/4.1.82.Final/netty-handler-4.1.82.Final.jar

Unpacking io/netty/netty-resolver/4.1.82.Final/netty-resolver-4.1.82.Final.jar (libraries:io.netty:netty-resolver:4.1.82.Final) to libraries/io/netty/netty-resolver/4.1.82.Final/netty-resolver-4.1.82.Final.jar

Unpacking io/netty/netty-transport/4.1.82.Final/netty-transport-4.1.82.Final.jar (libraries:io.netty:netty-transport:4.1.82.Final) to libraries/io/netty/netty-transport/4.1.82.Final/netty-transport-4.1.82.Final.jar

Unpacking io/netty/netty-transport-classes-epoll/4.1.82.Final/netty-transport-classes-epoll-4.1.82.Final.jar (libraries:io.netty:netty-transport-classes-epoll:4.1.82.Final) to libraries/io/netty/netty-transport-classes-epoll/4.1.82.Final/netty-transport-classes-epoll-4.1.82.Final.jar

Unpacking io/netty/netty-transport-native-epoll/4.1.82.Final/netty-transport-native-epoll-4.1.82.Final-linux-x86_64.jar (libraries:io.netty:netty-transport-native-epoll:4.1.82.Final:linux-x86_64) to libraries/io/netty/netty-transport-native-epoll/4.1.82.Final/netty-transport-native-epoll-4.1.82.Final-linux-x86_64.jar

Unpacking io/netty/netty-transport-native-epoll/4.1.82.Final/netty-transport-native-epoll-4.1.82.Final-linux-aarch_64.jar (libraries:io.netty:netty-transport-native-epoll:4.1.82.Final:linux-aarch_64) to libraries/io/netty/netty-transport-native-epoll/4.1.82.Final/netty-transport-native-epoll-4.1.82.Final-linux-aarch_64.jar

Unpacking io/netty/netty-transport-native-unix-common/4.1.82.Final/netty-transport-native-unix-common-4.1.82.Final.jar (libraries:io.netty:netty-transport-native-unix-common:4.1.82.Final) to libraries/io/netty/netty-transport-native-unix-common/4.1.82.Final/netty-transport-native-unix-common-4.1.82.Final.jar

Unpacking it/unimi/dsi/fastutil/8.5.9/fastutil-8.5.9.jar (libraries:it.unimi.dsi:fastutil:8.5.9) to libraries/it/unimi/dsi/fastutil/8.5.9/fastutil-8.5.9.jar

Unpacking net/java/dev/jna/jna/5.12.1/jna-5.12.1.jar (libraries:net.java.dev.jna:jna:5.12.1) to libraries/net/java/dev/jna/jna/5.12.1/jna-5.12.1.jar

Unpacking net/java/dev/jna/jna-platform/5.12.1/jna-platform-5.12.1.jar (libraries:net.java.dev.jna:jna-platform:5.12.1) to libraries/net/java/dev/jna/jna-platform/5.12.1/jna-platform-5.12.1.jar

Unpacking net/sf/jopt-simple/jopt-simple/5.0.4/jopt-simple-5.0.4.jar (libraries:net.sf.jopt-simple:jopt-simple:5.0.4) to libraries/net/sf/jopt-simple/jopt-simple/5.0.4/jopt-simple-5.0.4.jar

Unpacking org/apache/commons/commons-lang3/3.12.0/commons-lang3-3.12.0.jar (libraries:org.apache.commons:commons-lang3:3.12.0) to libraries/org/apache/commons/commons-lang3/3.12.0/commons-lang3-3.12.0.jar

Unpacking org/apache/logging/log4j/log4j-api/2.19.0/log4j-api-2.19.0.jar (libraries:org.apache.logging.log4j:log4j-api:2.19.0) to libraries/org/apache/logging/log4j/log4j-api/2.19.0/log4j-api-2.19.0.jar

Unpacking org/apache/logging/log4j/log4j-core/2.19.0/log4j-core-2.19.0.jar (libraries:org.apache.logging.log4j:log4j-core:2.19.0) to libraries/org/apache/logging/log4j/log4j-core/2.19.0/log4j-core-2.19.0.jar

Unpacking org/apache/logging/log4j/log4j-slf4j2-impl/2.19.0/log4j-slf4j2-impl-2.19.0.jar (libraries:org.apache.logging.log4j:log4j-slf4j2-impl:2.19.0) to libraries/org/apache/logging/log4j/log4j-slf4j2-impl/2.19.0/log4j-slf4j2-impl-2.19.0.jar

Unpacking org/joml/joml/1.10.5/joml-1.10.5.jar (libraries:org.joml:joml:1.10.5) to libraries/org/joml/joml/1.10.5/joml-1.10.5.jar

Unpacking org/slf4j/slf4j-api/2.0.1/slf4j-api-2.0.1.jar (libraries:org.slf4j:slf4j-api:2.0.1) to libraries/org/slf4j/slf4j-api/2.0.1/slf4j-api-2.0.1.jar

Starting net.fabricmc.loader.impl.game.minecraft.BundlerClassPathCapture

[16:14:21] [INFO] [FabricLoader/GameProvider]: Loading Minecraft 1.20.1 with Fabric Loader 0.15.3

[16:14:22] [INFO] [FabricLoader/GameRemap]: Fabric is preparing JARs on first launch, this may take a few seconds...

[16:14:24] [ERROR] [FabricLoader/]: Uncaught exception in thread "main"

java.lang.RuntimeException: An exception occurred when launching the server!

	at net.fabricmc.loader.impl.launch.server.FabricServerLauncher.main(FabricServerLauncher.java:71)

	at net.fabricmc.installer.ServerLauncher.main(ServerLauncher.java:69)

Caused by: java.lang.RuntimeException: error remapping game jars [/data/versions/1.20.1/server-1.20.1.jar]

	at net.fabricmc.loader.impl.game.GameProviderHelper.deobfuscate(GameProviderHelper.java:244)

	at net.fabricmc.loader.impl.game.minecraft.MinecraftGameProvider.initialize(MinecraftGameProvider.java:321)

	at net.fabricmc.loader.impl.launch.knot.Knot.init(Knot.java:140)

	at net.fabricmc.loader.impl.launch.knot.Knot.launch(Knot.java:68)

	at net.fabricmc.loader.impl.launch.knot.KnotServer.main(KnotServer.java:23)

	at net.fabricmc.loader.impl.launch.server.FabricServerLauncher.main(FabricServerLauncher.java:69)

	... 1 more

Caused by: java.nio.file.FileSystemException: /data/./.fabric/remappedJars/minecraft-1.20.1-0.15.3/zipfstmp12607438164232900423.tmp: Operation not permitted

	at java.base/sun.nio.fs.UnixException.translateToIOException(Unknown Source)

	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(Unknown Source)

	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(Unknown Source)

	at java.base/sun.nio.fs.UnixFileAttributeViews$Posix.setMode(Unknown Source)

	at java.base/sun.nio.fs.UnixFileAttributeViews$Posix.setPermissions(Unknown Source)

	at java.base/java.nio.file.Files.setPosixFilePermissions(Unknown Source)

	at jdk.zipfs/jdk.nio.zipfs.ZipFileSystem.sync(Unknown Source)

	at jdk.zipfs/jdk.nio.zipfs.ZipFileSystem.lambda$close$9(Unknown Source)

	at java.base/java.security.AccessController.doPrivileged(Unknown Source)

	at jdk.zipfs/jdk.nio.zipfs.ZipFileSystem.close(Unknown Source)

	at net.fabricmc.loader.impl.lib.tinyremapper.OutputConsumerPath.close(OutputConsumerPath.java:226)

	at net.fabricmc.loader.impl.game.GameProviderHelper.deobfuscate0(GameProviderHelper.java:314)

	at net.fabricmc.loader.impl.game.GameProviderHelper.deobfuscate(GameProviderHelper.java:242)

	... 6 more

```

**Comment by itzg** [comment, marked as answer]
Sounds like you might have file ownership/permission mismatch in your `/mnt/network_storage/Minecraft_Worlds_Java/hc_tester` directory. The container will run the Minecraft process with uid 1000.

**Reply by Frankenberrypi** [reply]
That does seem to be the problem.  I tried setting uid=0 and did not get the error.  I don't think running as root is the correct answer, but at least I know I have to go chase down mount permission issues now.  Thanks for the help (and all the great work on this project).

**Reply by paul-fornage** [reply]
My system does not have the option to select UID 0, but I have given full control to everything I can and it doesn't work. The server is currently running with UID 3141 GUID 568, and the directory has full control for U-568, G-568, U-3000, U-3141, and U-1000. I am running this on TrueNas Scale with auto curseforge. Forge modpacks work fine, but this pack won't. I understand you probably won't be able to help, but if there is any specific advice, that would be greatly appreciated.

**Reply by itzg** [reply]
@paulwrath1223 just in case I can determine anything, what is an example of one that works and one that doesn't work? In general though it might take manually looking at any ownership user or group mismatches of subdirectories.

**Reply by paul-fornage** [reply]
Thanks for the reply! 'Chosen's modded adventure' (EDIT: Forge) and 'ATM9' (EDIT: Forge) work great, but 'Prominence II' (EDIT: Fabric) yields this error. I have checked all folders and the exact file it is trying to access, and they have the same permissions. I don't have access to the directory anymore because I'm using it to host the new working pack, but it was some temp file in the `.fabric` directory. Curiously, it was able to create and modify other files (including create and populate the problematic file), so I'm wondering if it spawns Fabric with a separate UID to Minecraft? I don't know how minecraft modding works, so IDEK if they are separate processes.

**Reply by itzg** [reply]
It all runs within the same process. It sounds like a particular mod in the failing modpack is perhaps the problem.

**Reply by paul-fornage** [reply]
Thanks for the response. I ended up just using a different pack

**Reply by paul-fornage** [reply]
So it turns out I ran into the problem again and found this thread looking up the problem. I can now confirm that it won't load Fabric by itself with no mods. I have given full permission to UID (and GUID) 568 which is what the container runs as in truenas, as well as UID 1000 which you mentioned is what the container assigns minecraft to.

**Reply by paul-fornage** [reply]
Here are the permissions for the exact file that was failed to read: 
![image](https://github.com/user-attachments/assets/00a73829-9578-42b3-b37a-5858dfbb4657)

The logs also seem to indicate that it is trying to set new permissions (`at java.base/sun.nio.fs.UnixFileAttributeViews$Posix.setPermissions(Unknown Source)`)

`ls -l` of the offending file from the container shell: `-rwxrwxrwx 1 minecraft minecraft 15453058 Aug 10 16:21 zipfstmp10884944103339798190.tmp`

**Reply by Paul-Preteau** [reply]
Did you find a solution? I am running into the same issue.

**Reply by MagikEh** [reply]
Just ran into this thread while hunting the same issue as @paul-fornage. 

Found several related issues on the FabricMC GitHub and TrueNAS Forum asking for help as well:
- https://github.com/orgs/FabricMC/discussions/2406
- https://github.com/orgs/FabricMC/discussions/4472
- https://forums.truenas.com/t/issue-launching-modded-minecraft-server/36471/4
- https://forums.truenas.com/t/setup-for-cobbleverse-server/47563/5

### My Setup
- Truenas Scale `ElectricEel-24.10.2.4`
- Minecraft app 1.13.15 (upgrading from... 1.8.xx or something
  - Configured as `Type: Fabric`
- Minecraft `1.21.8` (upgrading from `1.21.1`)

### Steps to bandaid-fix
1. Run the app/container startup script
2. Rename the file from `zipfstmp....tmp` to `server-intermediary.jar`
   - If you have access to the dir as a mount, that's likely easiest.
   - If you need to do it via the TrueNAS >  System > Shell, use the following:
     ```sh
     # DON'T RUN SHELL COMMANDS FROM RANDOM PEOPLE ON THE INTERNET, use tools like https://explainshell.com to first understand what randos are asking you to run on your system.

     sudo -s
     # Enter your password as prompted.
     cd /mnt/${MY_POOL_NAME}/${MY_DATASET_NAME}/path/to/app/persistent/volume/.fabric/remappedJars/minecraft-
     ls -haltr
     # Take note of the bottom-most file for the next command
     mv zipfstmp$NUMBERS.*tmp server-intermediary.jar
     ```
3. Re-run the app/container startup script and it'll boot.


#### Rationale
Error in the logs show specifically:
```
at java.base/sun.nio.fs.UnixFileAttributeViews$Posix.setMode(Unknown Source)
at java.base/sun.nio.fs.UnixFileAttributeViews$Posix.setPermissions(Unknown Source)
at java.base/java.nio.file.Files.setPosixFilePermissions(Unknown Source)
at jdk.zipfs/jdk.nio.zipfs.ZipFileSystem.sync(Unknown Source)
...
at net.fabricmc.loader.impl.lib.tinyremapper.FileSystemReference.close(FileSystemReference.java:131)
at net.fabricmc.loader.impl.lib.tinyremapper.OutputConsumerPath.close(OutputConsumerPath.java:213)
at net.fabricmc.loader.impl.game.GameProviderHelper.deobfuscate0(GameProviderHelper.java:314)
at net.fabricmc.loader.impl.game.GameProviderHelper.deobfuscate(GameProviderHelper.java:242)
```
After having a quick peek at [the forge-loader source](https://github.com/FabricMC/fabric-loader/blob/master/src/main/java/net/fabricmc/loader/impl/game/GameProviderHelper.java#L369) for specifically the line numbers that were giving issue, I realized it was occurring when the loader had completed it's work and attempting to rename and re-perm the temp file.

I ran the loader a few times and observed that it was creating `tmp` files as the root user, rather than the user that it was set to be running the app as, and when attempting to chown them to /something/ it was failing:
```
cd ./.fabric/remappedJars/minecraft-1.21.8-0.17.2
ls -l
> total 69121
> drwxrwx--- 2 MagikEh        6 Sep 22 01:00 ./
> drwxrwx--- 4 MagikEh        4 Sep 21 23:42 ../
> -rwxrwx--- 1 MagikEh 19368888 Sep 22 00:53 zipfstmp11820519387640126180.tmp*
> -rwxrwx--- 1 MagikEh 19368888 Sep 22 00:44 zipfstmp14931245180372308889.tmp*
> -rwxrwx--- 1 root    19368888 Sep 22 01:00 zipfstmp1953920020336519961.tmp*    #<--- Newly created file is root user again :C
> -rwxrwx--- 1 MagikEh 19368888 Sep 21 23:42 zipfstmp2251717948150502019.tmp*
```
I then tried straight up `chown`ing the files to all be the runtime user (`MagikEh` in this case from a previous 1.21.1 working fileset) and when re-running the loader, it would again, just make a new `tmp` file rather than reading in the existing one(s). (Oh wow, would you  look at those 19 trailing random chars that're likely to prevent tmp file clobbering >.> )

So given the loader was just making new tmp files of /exactlly/ the same size, I figured I could try to force the rename, and that's how I got past that.

### Musings
I think... this may have something to do with how the docker container is spun up as a 'TrueNAS app' in that there are two distinct steps of first creating the files on the filesystem as a privileged user, then actually running the app. I'm no expert, but would love some insight, if possible, from the ones putting together [the TrueNAS app](https://github.com/truenas/apps/tree/master/trains/community/minecraft) (@sonicaj)?

**Reply by MagikEh** [reply]
Follow up observations on the TrueNAS app to maybe get more discussion going on what the root cause of this is (and why it was working fine in the older MC 1.21.1 versions of Fabric):

The app's user is defined to run as the default 568 app user:
<img width="229" height="221" alt="image" src="https://github.com/user-attachments/assets/e40a2688-4cd0-4c60-82e7-f9f4834be734" />

Yet the app's 
- container is `idfk:unknown`
- runtime context is `root:root` 
<img width="661" height="1099" alt="image" src="https://github.com/user-attachments/assets/a11f0d73-77fc-45c6-bb80-79bd040b6c3a" />

**Reply by Teufelauto** [reply]
@MagikEh, you are a legend. I have a similar enough TrueNAS 25.04.2.4 with a similar problem, and this 'one-simple-trick' works a treat, and is _actually_ simple.
> ### Steps to bandaid-fix
> 1. Run the app/container startup script
> 2. Rename the file from `zipfstmp....tmp` to `server-intermediary.jar`
>    
>    * If you have access to the dir as a mount, that's likely easiest.

Dog-gone right, it's easiest. I've been struggling for weeks, and renaming one file in SMB folder gets me up and running again. I tried just about every ACL permutation I could think of. Thanks for directing me here from the TrueNAS forums.

**Reply by MagikEh** [reply]
Glad to help! Now if only we can find the underlying issue that actually needs to be corrected xD

**Reply by Teufelauto** [reply]
I have Fabric installing as intended on TrueNAS 25.04.2.4 Fangtooth with the following procedure- **The key is creating the intended 'Host Path' dataset with the _APPS Preset_, and not changing its Role.**

I believe the root of our problem is with TrueNAS's 'Dataset Presets' creation. If the dataset does not start life with an APP Preset, it won't work, and can't be made to work, as far as I can tell. May be a bug (_maaaybe_ an intended quirk) in TrueNAS Roles.

1. So, I have `/mnt/minepool/Minecraft-Java-Servers` as an SMB Dataset with root owner, apps user/group, and my SMB user. (builtin_admin and builtin_users groups are automatically added) 'Apply Permissions Recursively', and 'Apply Permissions to Child Datasets' need to be checked in the ACL.

2. In the above parent, I create a dataset for each server I'm hosting using the **_APPS Preset_**. If it asks to set ACL, hit the _Return to Pool List_ button. It will still inherit the users and allow SMB access. `/mnt/minepool/Minecraft-Java-Servers/test-server` can be accessed by Windows File Explorer through the parent. No need to mess with this dataset's ACL or changing 'Roles'.

3. Minecraft App can now be installed using the above `test-server` dataset as the Minecraft Data Storage mount path.

I narrowed it down to the Preset, by successfully installing Minecraft Data Storage in an IXVolume. I figured that must be an APP type dataset, so I made a test dataset with great success! I don't know what makes the APPS Preset special. I was installing the app with a dedicated 'minecraftuser' before, but now, I'm just going to use the apps default user.

I hope this helps, being documented here. Easier creating the datasets this way, than having to rename an obscure file. Either way, this note is going in my own Minecraft Server Installation Procedures document. Cheers!
